### PR TITLE
InfluxDB: Refactor annotation transform logic

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

1. Refactoring InfluxDB annotation transform logic with typed variables. 
2. Setting default retention policy when there is no retention policy in query model (you just add a new query row)
3. Run annotation query earlier in the chain to support both backend and frontend mode.

Part of Epic: https://github.com/grafana/grafana/issues/65045

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by gtk-grafana** [comment]
Getting errors on this branch I'm not seeing on play or main
image
image

Running a query of `SELECT * FROM "cpu"` works in the panel, but throws an error in the annotation query editor
image

**Comment by itsmylife** [comment]
@bohandley I cannot reproduce the retention policy issue. Can you please provide exact steps with data source info (v1.x or 2.x), explore or dashboard, etc?

**Comment by bohandley** [comment]
Here are the steps I did to get the retention policy error:

1. Checkout branch ismail/influxdb-refactoring-7
2. Run `make devenv sources=influxdb` in the terminal to start up influxdb instance (this is from influxdb:latest, InfluxDB v2.7.1)
3. Create a new dashboard.
4. Add a visualization.
5. Select the data source `gdev-influxdb-influxql` (this is from influxdb:latest, InfluxDB v2.7.1).
6. Switch to code editor and write the query `select * from cpu`, see the query works and returns series.
7. Click 'apply'
8. Navigate to dashboard settings, select annotations and click add annotation query.
9. Choose `gdev-influxdb-influxql` as the data source.
10. Enter `select * from cpu` in the input for InfluxQL query.
11. Click outside the input.
12. See the error `runRequest.catchError {"message":"InfluxDB Error: retention policy not found: undefined","res":{"data":[],"state":"Error","error":{"refId":"A","message":"retention policy not found: undefined","status":500},"errors":[{"refId":"A","message":"retention policy not found: undefined","status":500}]}}`

**Comment by bohandley** [comment]
I had the feature toggle on, `influxdbBackendMigration = true`. With it off, I do not get the retention policy error.

**Comment by itsmylife** [comment]
@bohandley Retention policy error is fine since I did not touch the code editor yet. There are some issues with it (some of them existing ones). But I will fix this one in this PR to have a clear state. btw `select * from cpu` returns `~26mb` of data. At least for testing, you might want to use this query `SELECT mean("usage_idle") FROM "default"."cpu" WHERE $timeFilter GROUP BY time($__interval) fill(null)`

File: public/app/plugins/datasource/influxdb/components/editor/query/influxql/visual/VisualInfluxQLEditor.tsx [hunk 1]
```diff
@@ -54,6 +54,16 @@ export const VisualInfluxQLEditor = (props: Props): JSX.Element => {
...
+  useEffect(() => {
+    if (!policy) {
+      props.onChange({
+        ...query,
+        policy: retentionPolicies[0],
+      });
+      props.onRunQuery();
+    }
+  }, [policy, props, query, retentionPolicies]);
+
```
Comment (itsmylife) [inline public/app/plugins/datasource/influxdb/components/editor/query/influxql/visual/VisualInfluxQLEditor.tsx:66]: If there is no retention policy (you just add a new query row), set back that default policy into the query model

**Review by bohandley — APPROVED** [review]
Nice! Everything is working for me 🚀 
Screenshot 2023-06-13 at 11 42 43 AM
Screenshot 2023-06-13 at 11 44 45 AM

**Pull request merged**